### PR TITLE
image-builder: make DAX support optional

### LINKED ISSUE #246: fc: DAX support should be optional
Firecracker doesn't support NVDIMM hence DAX support in the image must be optional

### PR DESCRIPTION
**Pull request by devimc**
Not all hypervisor support NVDIMM hence DAX support MUST BE enabled explicitly
setting the DAX environment variable to 'yes'

fixes #246

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: image-builder/image_builder.sh [hunk 1]
```diff
@@ -39,8 +40,19 @@ Options:
 Extra environment variables:
 	AGENT_BIN:  use it to change the expected agent binary name
 	AGENT_INIT: use kata agent as init process
+	DAX: If 'yes' will build the image with DAX support. The resulting image
+	     *can't* be mounted because the first 2M are reserved for the device
```
Comment (jodh-intel) [inline image-builder/image_builder.sh]: Eeek - I think this is going to break https://github.com/kata-containers/runtime/blob/master/data/kata-collect-data.sh.in then unless we `losetup -o <offset> ...` ?

The problem is, how is `kata-collect-data.sh` going to know to apply the offset? We could add in an arch check but I'm starting to think we need a metadata file for each image (basically `osbuilder.yaml` + extra fields that would live *outside* the VM in the assets directory alongside the images.

**Comment by sboeuf** [comment]
@devimc I thought that our CI was using images directly from the packages?
My concern is that to handle both cases (VMM with DAX support vs VMM without DAX support), we will need two different images. And if we want two different images not to be built every time by our CI, then we will end up with two different images in our packages.
Is that a good summary of the situation here, or am I missing something?

**Comment by devimc** [comment]
@sboeuf 

> @devimc I thought that our CI was using images directly from the packages?
nop

you're right, we'll need two different images, one for firecracker and other for qemu+nvidimm, I don't see any other alternative

File: image-builder/image_builder.sh (hunk 1 above)
Comment (devimc) [reply image-builder/image_builder.sh]: my bad, after reading `losetup` documentation, now I see that it's possible to mount the image using `-o offset`

Comment (jodh-intel) [reply image-builder/image_builder.sh]: Yep - the problem is knowing when to use that option though.

**Comment by jodh-intel** [comment]
We probably want to bring this up at the Arch Committee meeting as there are a few implications here including:

- Packaging implications of course.
- Updating the `kata-collect-data.sh` script.
- Testing implications.
- Updating docs (Install docs [prolly not quite yet, but soon]? Dev guide? Architecture guide?)
- ...?

**Comment by GabyCT** [comment]
Tested this change locally and FC is not failing

**Comment by GabyCT** [comment]
I can run a container with FC without issues

**Comment by jodh-intel** [comment]
Restarted one of the Travis jobs which had timed out (the other passed)...

**Comment by Pennyzct** [comment]
@devimc @chavafg I think this pr will fix ARM CI broken issue. ;) already tested it on local machine.

File: image-builder/image_builder.sh [hunk 2]
```diff
@@ -39,8 +40,22 @@ Options:
...
+
+
+	When DAX is 'yes', the following diagram shows how a 128M image will looks like:
+		 .-----------------------------------.
+		 |-- 2 MB --|-------- 126 MB --------|
+		 | Metadata | Rootfs (/bin,/usr,etc) |
+		 '-----------------------------------'
+
+		 The resulting image can be mounted if the offset of 2 MB is specified:
+		 $ sudo losetup -v -fP -o $((2*1024*1024)) kata-containers.img
```
Comment (jcvenegas) [inline image-builder/image_builder.sh]: Is there a way for end users a way to verify the image has dax? or not ?  I know some people open the image directly to modify as needed and do some testing in Kata images.

File: image-builder/image_builder.sh [hunk 3]
```diff
@@ -276,6 +295,26 @@ create_rootfs_disk()
...
+{
+	dax_header_bytes=$((DAX_HEADER_SZ*1024*1024))
+	info "Set device namespace information (metadata)"
+	# Fill out namespace information
+	tmp_img="$(mktemp)"
+	chmod 0644 "${tmp_img}"
+	# metadate header
+	dd if=/dev/zero of="${tmp_img}" bs="${DAX_HEADER_SZ}M" count=1
+	# append image data (rootfs)
+	dd if="${IMAGE}" of="${tmp_img}" oflag=append conv=notrunc
```
Comment (jcvenegas) [inline image-builder/image_builder.sh]: just in case can you verify the image variable is not empty

**Comment by jcvenegas** [comment]
> We probably want to bring this up at the Arch Committee meeting as there are a few implications here including:
> 
> * Packaging implications of course.
> * Updating the `kata-collect-data.sh` script.
> * Testing implications.
> * Updating docs (Install docs [prolly not quite yet, but soon]? Dev guide? Architecture guide?)

Agree with @jodh-intel,  this support need more planning as will need to modify different things in our flow. @devimc  lets document the follow up changes to fully support this and also to talk about any possible issue somebody can point about .

**Comment by sboeuf** [comment]
@jcvenegas @devimc 
>Agree with @jodh-intel, this support need more planning as will need to modify different things in our flow. @devimc lets document the follow up changes to fully support this and also to talk about any possible issue somebody can point about

This would be great, please do so :)

**Pull request merged**